enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/readreplica/UpstreamDatabaseSelectionStrategy.java

```diff
@@ -17,11 +17,25 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program. If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.causalclustering.messaging.routing;
+package org.neo4j.causalclustering.readreplica;
 
+import org.neo4j.causalclustering.discovery.TopologyService;
 import org.neo4j.causalclustering.identity.MemberId;
+import org.neo4j.helpers.Service;
 
-public interface UpstreamDatabaseSelectionStrategy
+public abstract class UpstreamDatabaseSelectionStrategy extends Service
 {
-    MemberId upstreamDatabase() throws UpstreamDatabaseSelectionException;
+    protected TopologyService topologyService;
+
+    public UpstreamDatabaseSelectionStrategy( String key, String... altKeys )
```

Comment:
The way in which these keys work needs documentation. What separate between key and alternate keys?